install_files/ansible-base/roles/backup/files/0.3_restore.py

```diff
@@ -0,0 +1,163 @@
+import zipfile
+import subprocess
+import shutil
+from datetime import datetime
+from operator import itemgetter
+import calendar
+import traceback
+
+def replace_prefix(path, p1, p2):
+    """
+    Replace p1 in path with p2
+
+    >>> replace_prefix("/tmp/files/foo.bar", "/tmp", "/home/me")
+    "home/me/files/foo.bar"
+    """
+    common_prefix = os.path.commonprefix([path, p1])
+    if common_prefix:
+        assert path.find(common_prefix) == 0
+        # +1 so chop off the next path separator, which otherwise becomes a
+        # leading path separate and confuses os.path.join
+        path = path[len(common_prefix)+1:]
+    return os.path.join(p2, path)
+
+
+def extract_to_path(archive, member, path):
+    """
+    Extract from the zip archive `archive` the member `member` and write it to
+    `path`, preserving file metadata.
+    """
+    # Create all upper directories if necessary
+    upperdirs = os.path.dirname(path)
+    if upperdirs and not os.path.exists(upperdirs):
+        os.makedirs(upperdirs)
+
+    with archive.open(member) as source, \
+         file(path, "wb") as target:
+        shutil.copyfileobj(source, target)
+
+    # Update the timestamps as well (as best we can, thanks, conversion to
+    # localtime). This only actually works if the .zip was created on a
+    # machine where the timezone was set to UTC, but it might be good
+    # enough since we just need the relative order of timestamps (they will
+    # all be normalized anyway).
+    if hasattr(member, 'date_time'):
+        timestamp = calendar.timegm(member.date_time)
+        os.utime(path, (timestamp, timestamp))
+
+def restore_config_file(zf):
+    print "* Migrating SecureDrop config file from backup..."
+
+    # Restore the original config file
+    for zi in zf.infolist():
+        if "var/www/securedrop/config.py" in zi.filename:
+            extract_to_path(zf, "var/www/securedrop/config.py", "/var/www/securedrop/config.py")
+
+    print "chowning config"
+
+    subprocess.call(['chown', '-R', 'www-data:www-data', "/var/www/securedrop/config.py"])
+
+def restore_securedrop_root(zf):
+    print "* Migrating directories from SECUREDROP_ROOT..."
+
+    # Restore the original source directories and key files
+    for zi in zf.infolist():
+        if "var/lib/securedrop/store" in zi.filename:
+            extract_to_path(zf, zi, replace_prefix(zi.filename,
+                "var/lib/securedrop/store", "/var/lib/securedrop/store"))
+        elif "var/lib/securedrop/keys" in zi.filename:
+            # TODO: is it a bad idea to migrate the random_seed from the
+            # previous installation?
+            extract_to_path(zf, zi, replace_prefix(zi.filename,
+                "var/lib/securedrop/keys", "/var/lib/securedrop/keys"))
+
+    subprocess.call(['chown', '-R', 'www-data:www-data', "/var/lib/securedrop"])
+
+
+def restore_database(zf):
+    print "* Migrating database..."
+
+    extract_to_path(zf, "var/lib/securedrop/db.sqlite", "db.old.sqlite")
```

Comment:
Does this extract_to_path call extract to the correct location? Shouldn't it extract to "/var/lib/securedrop/db.sqlite"?